### PR TITLE
Make minted usable with Pygments stylenames containing digits

### LINKED ISSUE #210: Numbers in styles
Using styles with numbers in their names does not work, eg. styles from [base16-pygments](https://github.com/idleberg/base16-pygments).
So
```latex
\usemintedstyle{base16-classic-light}
```
won't work, default style is used instead.

Actually one can tweak the *.py files in order to use roman style numbering, but it's kind of hacky.

### LINKED ISSUE #294: Build failure with Pygments 2.8.0 due to it stylesheet being loaded with character `-` having catcode letter
Perhaps the relevant code lines are
https://github.com/gpoore/minted/blob/5d72859d714a6f2f6a42eec524476994d954b960/source/minted.dtx#L2448-L2458

If the reason is that the `-` character is used in some control sequence name internal to minted, it can always be used inside  `\@nameuse` with its normal catcode 12.

As Pygments 2.8.0 stylesheet for LaTeX contains a length assignment which uses `-`, it breaks due to the wrong catcode setting.

MWE: the one of first linked issued below.

Related:

https://github.com/pygments/pygments/issues/1734

https://github.com/pygments/pygments/issues/1731

### LINKED ISSUE #299: custom pygments style detected but ignored only in minted
Hi,

I'm running an up-to-date debian unstable. I installed [base16-pygments](https://github.com/mohd-akram/base16-pygments) by running `pip3 install pygments-base16`.

When I try the following: `xelatex -shell-escape -halt-on-error main.tex` on this file :

```tex
\documentclass[12pt]{article}

\usepackage{minted}
\usemintedstyle{base16-onedark}

\begin{document}
  \mintinline{ocaml}{let f x = Some x}
\end{document}
```

It produces a file with default minted color scheme. It doesn't use the specified `base16-onedark`.

If I change the name to something that doesn't exist like `fjesfkljesklfjklesf`, I have a proper error:

```
(./_minted-main/default-pyg-prefix.pygstyle)Could not find style module 'fjesfkljesklfjklesf'.
system returned with code 256
```

So, minted finds that my style exists but just seems to ignore it. So I asked myself : maybe the style is wrong ?

I tried to call pygments myself, using: `pygmentize -f html -O full,style=base16-onedark -l ocaml -o out.html code.ml` on this file:

```ocaml
let f x = Some x
```

I checked the result with `xdg-open out.html` and the color scheme is the right one.

OK, maybe minted is just ignoring all my styles ? So I changed the `\usemintedstyle{base16-onedark}` to `\usemintedstyle{monokai}` (monokai is installed by default with pygment on my system) and this one is working.

So for some reasons, minted knows that `base16-onedark` exists (otherwise it would fail saying it can't find it) but ignores it completely and I can't see why...

Everything works fine with pygmentize so I believe the issue comes from minted. Do you have any clue ?

Cheers.

### PR DESCRIPTION
**Pull request by jfbu**
Fix #210 

Close #299

Close #294 (which had already a work-around done at Pygments level; the work-around will need no adjustment although it will become superfluous)

Relates (and resolves purely on minted side)
https://github.com/pygments/pygments/issues/1770

I have not tested this a lot and my knowledge of minted codebase is partial.

Some comments: if I did not make some logical error, the thing is robust with respect to an evolving document from which code snippets for various styles are added or deleted. The aux file is used.

 *obsoleted by later commits*: <strike>It will contain redundant data, for example if only the default Pygments style is used, each code snippet creates a line in the aux file, identical to the others. But this is not important in my view. On the other hand,</strike> 

the code tries to be economical with calls to pygmentize. Normally the style file for a given style name will be queried only once. 

*Next sentence was wrong because I did not know then about `minted@cleancache`*: <strike>The cache repertory will keep on containing possibly not used anymore pygstyle files, if user deletes corresponding code snippets in source.</strike> 

If the aux file is deleted all style files will be queried again from pygmentize as the command prefixes then are not known and need to be reset.

With little modification this will allow to support Pygments style names with characters which are for example "active" in LaTeX. One needs some `\detokenize` in various places. But I kept the matter simple.

I tested using this file

```latex
\documentclass{article}
\usepackage[]{minted}

\begin{document}

\begin{minted}[style=stata-light]{python}
foo="-"  
\end{minted}

\begin{minted}[style=base16-onedark]{python}
foobase16="-"  
\end{minted}

\begin{minted}[style=algol_nu]{python}
fooalgol="-"  
\end{minted}

\begin{minted}[style=paraiso-dark]{python}
fooparaiso="-"  
\end{minted}

\end{document}
```

This test file was used with various additions, deletions, and permutations of the ordering of the code snippets to confirm robustness of the PR towards an evolving document source.  Also tested was manual removal of `.aux` file keeping cache. I also briefly tested with caching off, or frozencache. 

Succesful compilation confirmed it worked with digits and with the  `_` and `-`, with no catcode adjustments. So the upstream Pygments patch https://github.com/pygments/pygments/pull/1735 which worked around minted using `-` with catcode letter is now made superfluous (but does not harm).

### CODE REVIEW

File: source/minted.dtx [hunk 1]
```diff
@@ -2405,77 +2405,116 @@
...
-        \minted@addcachefile{\ifstrempty{#1}{default-pyg-prefix}{#1}.pygstyle}}%
+      \begingroup
+      \let\def\gdef
+      \catcode\string``=12
+      \catcode`\%=14
+      \endlinechar=-1\relax
+      \minted@input{%
+        \minted@outputdir\minted@cachedir/\minted@checkedstyle.pygstyle}%
+      \endgroup
+      \expandafter\let\csname minted@isloaded@\minted@checkedstyle\endcsname\@empty
```
Comment (jfbu) [inline source/minted.dtx:2495]: This probably needs a `\global` for safety/coherence

File: source/minted.dtx [hunk 2]
```diff
@@ -2405,77 +2405,116 @@
...
         \let\def\gdef
         \catcode\string``=12
-        \catcode`\_=11
-        \catcode`\-=11
         \catcode`\%=14
         \endlinechar=-1\relax
         \minted@input{\minted@outputdir\minted@jobname.out.pyg}%
-        \endgroup}%
+        \endgroup
+        \expandafter\let\csname minted@isloaded@\minted@checkedstyle\endcsname\@empty
```
Comment (jfbu) [inline source/minted.dtx:2516]: same here

File: source/minted.dtx [hunk 3]
```diff
@@ -2405,77 +2405,116 @@
...
+% We have to do some tricks with |\endlinechar| to prevent |\input| from inserting unwanted whitespace.  That is primarily for inline commands, where it would introduce a line break.  There is also the very unorthodox |\let\def\gdef| to make sure that macros are defined globally.  The catcode for |%| must be changed to the default value in case \pkg{minted} is being used in a |.dtx| file; otherwise, \LaTeX\ comments in the Pygments style files will appear as literal text.
 %
 % If a style is not given, then revert to the |default| style, but create macros with prefix |PYG|, and create |default-pyg-prefix.pygstyle| if caching is on.  This allows a graceful fallback in the event that style is empty.  It is also purposefully used to create a complete set of macros with prefix |PYG|, so that the symbol macros may be used, as described next.
```
Comment (jfbu) [inline source/minted.dtx:2429]: This sentence needs update as now the name of the default style file is `default.pygstyle` rather.

**Comment by jfbu** [comment]
I accidentally added to the index the binary file `minted.pdf` when setting up this PR. I will thus force-push to remove it. I guess the reason for not adding `minted.pdf` to some `.gitignore` is that it is indeed under version control,  but only committed for major releases? Force-pushed commit will incorporate some further documentation adjustments (as reviewed above).

**Comment by jfbu** [comment]
Sorry for latest force-pushed commit.

This is ready for review.

**Comment by jfbu** [comment]
The latex `.aux` file after addition of a new code snippet will stabilize only after two runs. I considered this not be an essential problem, for example if the document maintains a table of listings, each addition requires two runs for the pdf to stabilize. Here the pdf is ok on first run after addition of a code snippet, but the `.aux` file will stabilize only after two runs. Admittedly, when using `Latexmk` the changing `.aux` file will trigger a third compilation. I considered this as not really important.

However, while I have in mind the details, it is probably better I add a commit to let the `.aux` file acquire final shape via the first post-changed-source latex run, not the second. I will do this soon.  So ready-for-review was not quite true...

**Comment by jfbu** [comment]
again (or rather at last) ready for review...

**Comment by gpoore** [comment]
Thanks for working on this! A counter-based naming scheme can definitely work around the naming issues.

The `.aux` changes are acceptable if necessary but not ideal...it would be nice to minimize compiles. In looking at these issues again, I had an idea for a completely different approach that may be possible with minimal changes and no `.aux`. I'm going to see if that works, otherwise I will merge your solution. I finally have a few days when I can work on `minted` again, so one way or another we'll get this and some other issues resolved, and a new version released.

**Comment by gpoore** [comment]
I've just committed an alternative solution for this. If you find any issues with it, or disadvantages relative to your solution, let me know and we can go from there.

The basic idea is to create all style definitions using a `PYG` command prefix. This would cause the different styles to conflict/overwrite each other, so style macros are only ever defined locally within `minted` commands/environments.

Each style file is read into a global macro named after the style name (so files are only ever read once). These global macros containing definitions are used within commands/environments to define the appropriate styles locally. This does add a little macro definition overhead compared to your solution, but that should be minimal. Style files are only read once, so I think that should avoid the main potential for overhead with redefinitions. Because all styles always use the same command prefix and naming format, it is possible for multiple documents to share a common cache, which could be difficult with a counter-based approach.

One disadvantage is that since styles are only defined locally, it isn't as straightforward to customize them via additional macro programming. I suppose it would be possible to add a new option for doing this, maybe something like `custompyg` that would specify commands which are only used after styles are defined within commands/environments.

**Comment by jfbu** [comment]
> I've just committed an alternative solution for this. If you find any issues with it, or disadvantages relative to your solution, let me know and we can go from there.

I have only brief comments, keeping the general ones here, made the others at the commit.

> 
> The basic idea is to create all style definitions using a `PYG` command prefix. This would cause the different styles to conflict/overwrite each other, so style macros are only ever defined locally within `minted` commands/environments.

> 
> Each style file is read into a global macro named after the style name (so files are only ever read once). These global macros containing definitions are used within commands/environments to define the appropriate styles locally. This does add a little macro definition overhead compared to your solution, but that should be minimal. Style files are only read once, so I think that should avoid the main potential for overhead with redefinitions. Because all styles always use the same command prefix and naming format, it is possible for multiple documents to share a common cache, which could be difficult with a counter-based approach.
> 

For the approach I proposed it would only be a matter of directing the data it stores in the `.aux` file into a file contained in the cache, and using it from there.

By the way the current mechanism of executing `\minted@cleancache` seems to make using same cache for multiple documents sub-optimal no? the "old" contents listing of cached style files is stored in `.aux` file, so when a document does not need a style, it deletes it from cache, right? Or is there some option for sharing cache and avoiding this (maybe I should read the docs...)

> One disadvantage is that since styles are only defined locally, it isn't as straightforward to customize them via additional macro programming. I suppose it would be possible to add a new option for doing this, maybe something like `custompyg` that would specify commands which are only used after styles are defined within commands/environments.

I think your approach is sound. It has the drawback you mentioned of each inline or display minted code snippet to cause the Pygments style definitions be all re-executed (even twice when a style is first encountered). This is probably small overhead (btw seemingly simple LaTeX instruction such as `\input` generated at some point tens of thousands of macro expansion when handling long file names like those with two sha hashes as done by minted, and although this has been improved in recent LaTeX, the cost of executing a few dozen TeX macro definitions will become relatively noticeable only if done hundreds of times).

It has the potential advantage of easing up sharing the cache. As I commented the approach of this PR could be extended by simply storing in the cache the style-to-prefix mapping rather than storing it in the `.aux` file. It may be a bit perturbing that the state of the cache, i.e. not only the actual style files it contains but also the command prefixes use therein will depend on the history of the document, or multiple documents sharing the same cache, depending on the history of additions/deletions of usages of various styles, but personally I ended up considering this to be ok feature.

Your approach also has the advantage of not worrying about the TeX escape macros, as thay are always defined with common `\PYG` prefix, and there is no need to query Pygments default style systematically, hence some economy here.

To end on a light note, I did not actually test your approach, so you are free to keep going at your own risks and perils ;-)

**Comment by gpoore** [comment]
Regarding cache: There are no special features for sharing a cache, so there will be some inefficiency with a shared cache. A document can delete a cache file that it no longer needs, but that is needed by another document, in which case the file must be regenerated. In practice, this is often rare enough to be insignificant. You are correct that saving macro name data in the cache rather than the `.aux` would make your approach work with a shared cache.

I think I will go with my solution for now, mostly because I think that the common `\PYG` prefix will simplify things for the future. Thanks for your work with this! Even though the code from your pull request may not be used, it has been very helpful in thinking through possible solutions and tradeoffs. Also, thanks again for your comments on other parts of `minted` elsewhere.

**Pull request closed without merging**